### PR TITLE
Support TrueType hinting

### PR DESCRIPTION
**Pull request by jenskutilek**
See #334.

### CODE REVIEW

**Comment by jenskutilek** [comment]
I have switched the TrueType compilation from htic code to FontTools TTX assembly code.

It uses the TrueType data from the UFO lib keys as defined in https://github.com/unified-font-object/ufo-spec/pull/94.

**Comment by jenskutilek** [comment]
Builds will fail until https://github.com/fonttools/fonttools/pull/2005 is merged.

File: Lib/ufo2ft/postProcessor.py [hunk 1]
```diff
@@ -113,6 +113,10 @@ def process(
 
         if compileTrueTypeHinting and "glyf" in self.otf:
             self._compile_truetype_hinting()
+            # Force compilation of the font to avoid glyph name problems
```
Comment (jenskutilek) [inline Lib/ufo2ft/postProcessor.py:116]: Here I had to resort to a kind of hack ... forcing the font to be compiled once by saving it and opening it again, otherwise the glyph names go out of sync and glyphs will disappear from the font.

File: Lib/ufo2ft/postProcessor.py [hunk 2]
```diff
@@ -113,6 +113,10 @@ def process(
...
+            tmp = BytesIO()
+            self.otf.save(tmp)
+            self.otf = TTFont(tmp)
```
Comment (jenskutilek) [inline Lib/ufo2ft/postProcessor.py:119]: Here I had to resort to a kind of hack ... forcing the font to be compiled once by saving it and opening it again, otherwise the glyph names go out of sync and glyphs will disappear from the font.

**Comment by jenskutilek** [comment]
I think this PR is good to go. Any objections, comments, etc.? Should we wait unti the UFO spec has been officially updated?

**Comment by jenskutilek** [comment]
As the instructions are now part of the UFO spec, we may consider merging this.
Related merged pull request: https://github.com/unified-font-object/ufo-spec/pull/94

**Comment by moyogo** [comment]
@jenskutilek if you could review https://github.com/unified-font-object/ufo-spec/pull/170 then we can update this PR for the flags.

File: Lib/ufo2ft/instructionCompiler.py [hunk 3]
```diff
@@ -0,0 +1,227 @@
...
+from fontTools.pens.hashPointPen import HashPointPen
+from fontTools.ttLib.tables._g_l_y_f import (
+    OVERLAP_COMPOUND,
+    ROUND_XY_TO_GRID,
+    USE_MY_METRICS,
+)
+
+logger = logging.getLogger(__name__)
+
+ufoLibKey = "public.truetype.instructions"
```
Comment (moyogo) [inline Lib/ufo2ft/instructionCompiler.py:15]: Add something like `TRUETYPE_INSTRUCTIONS_KEY = "public.truetype.instructions"`, `OBJECT_LIBS_KEY = "public.objectLib"` and `OBJECT_IDENTIFIERS_KEY = "public.objectIdentifiers"` to  constants.py and use them below.

File: Lib/ufo2ft/instructionCompiler.py [hunk 4]
```diff
@@ -0,0 +1,227 @@
...
+                return
+            asm = ttdata.get(key, None)
+            if asm is not None:
+                self.font[table_tag] = table = ttLib.newTable(table_tag)
+                table.program = ttLib.tables.ttProgram.Program()
+                table.program.fromAssembly(asm)
+                # Roundtrip once, or if the font is dumped to XML before having
+                # been saved, the assembly code if will look awful.
+                table.program._assemble()
+                table.program._disassemble(preserve=True)
```
Comment (moyogo) [inline Lib/ufo2ft/instructionCompiler.py:43]: This can be `glyf.program.fromBytecode(glyf.program.getBytecode())`.

File: Lib/ufo2ft/instructionCompiler.py [hunk 5]
```diff
@@ -0,0 +1,227 @@
...
+                        not in glyph.lib["public.objectLibs"]
+                        or ufo_component.identifier
+                        not in glyph.lib["public.objectLibs"][
+                            "public.objectIdentifiers"
+                        ]
+                        or "public.truetype.instructions"
+                        not in glyph.lib["public.objectLibs"][
+                            "public.objectIdentifiers"
+                        ][ufo_component.identifier]
+                    ):
```
Comment (moyogo) [inline Lib/ufo2ft/instructionCompiler.py:158]: This should use `glib.lib["public.objectLibs"][ufo_component.identifier]` directly.

File: Lib/ufo2ft/instructionCompiler.py [hunk 6]
```diff
@@ -0,0 +1,227 @@
...
+                        # c.flags &= ~ROUND_XY_TO_GRID
+                        # if c.x != 0 or c.y != 0:
+                        #     c.flags |= ROUND_XY_TO_GRID
+
+                        # Nope, don't change the flag:
+                        # https://github.com/googlefonts/ufo2ft/pull/425
+                        pass
+
+                        # "USE_MY_METRICS" has been set already by
+                        # outlineCompiler.OutlineTTFCompiler.autoUseMyMetrics
```
Comment (moyogo) [inline Lib/ufo2ft/instructionCompiler.py:173]: Maybe only link to the issue and remove the unused code.

File: Lib/ufo2ft/instructionCompiler.py [hunk 7]
```diff
@@ -0,0 +1,227 @@
...
+
+                    else:
+                        # Use values from lib
+
+                        flags = glyph.lib["public.objectLibs"][
+                            "public.objectIdentifiers"
+                        ][glyph.components[i].identifier][
+                            "public.truetype.instructions"
+                        ]
```
Comment (moyogo) [inline Lib/ufo2ft/instructionCompiler.py:182]: This should use `glib.lib["public.objectLibs"][glyph.components[i].identifier]` directly.

File: Lib/ufo2ft/instructionCompiler.py (hunk 3 above)
Comment (jenskutilek) [reply Lib/ufo2ft/instructionCompiler.py]: Thank you for your suggestions, I have made all changes accordingly.

File: Lib/ufo2ft/instructionCompiler.py [hunk 8]
```diff
@@ -0,0 +1,194 @@
+# -*- coding: utf-8 -*-
```
Comment (anthrotype) [inline Lib/ufo2ft/instructionCompiler.py:1]: we don't need this for python3, it's already utf-8

File: Lib/ufo2ft/instructionCompiler.py [hunk 9]
```diff
@@ -0,0 +1,194 @@
...
+        self.ufo = ufo
+        self.font = ttf
+
+    def _compile_program(self, key, table_tag):
+        assert table_tag in ("prep", "fpgm")
+        ttdata = self.ufo.lib.get(TRUETYPE_INSTRUCTIONS_KEY, None)
+        if ttdata:
+            formatVersion = ttdata.get("formatVersion", None)
+            if int(formatVersion) != 1:
+                logger.error(
```
Comment (anthrotype) [inline Lib/ufo2ft/instructionCompiler.py:33]: maybe consider raising a NotImplementedError instead of a logging.error which could easily be ignored, unless ignoring is desirable

File: Lib/ufo2ft/instructionCompiler.py [hunk 10]
```diff
@@ -0,0 +1,194 @@
...
+                    f"table '{table_tag}' will be empty in font."
+                )
+                return
+            asm = ttdata.get(key, None)
+            if asm is not None:
+                self.font[table_tag] = table = ttLib.newTable(table_tag)
+                table.program = ttLib.tables.ttProgram.Program()
+                table.program.fromAssembly(asm)
+                # Roundtrip once, or if the font is dumped to XML before having
+                # been saved, the assembly code if will look awful.
```
Comment (anthrotype) [inline Lib/ufo2ft/instructionCompiler.py:45]:
```suggestion
                # been saved, the assembly code will look awful.
```

how "awful"? this sounds like a bug in fonttools?

File: Lib/ufo2ft/postProcessor.py (hunk 2 above)
Comment (anthrotype) [reply Lib/ufo2ft/postProcessor.py:122]: why do we need to compile the hinting in the postProcessor, can't we do in the outlineCompiler itself?

File: Lib/ufo2ft/instructionCompiler.py [hunk 11]
```diff
@@ -0,0 +1,194 @@
...
+                value = ttdata.get(name, None)
+                if value is not None:
+                    setattr(maxp, name, value)
+
+        # Recalculate maxp.maxSizeOfInstructions
+        sizes = [
+            len(glyph.program.getBytecode())
+            for glyph in self.font["glyf"].glyphs.values()
+            if hasattr(glyph, "program")
+        ] + [0]
```
Comment (anthrotype) [inline Lib/ufo2ft/instructionCompiler.py:183]: instead of `+ [0]` use `max(sizes, default=0)`

File: Lib/ufo2ft/instructionCompiler.py [hunk 12]
```diff
@@ -0,0 +1,194 @@
...
+                                    f"component '{use_my_metrics_comp}' already."
+                                )
+                            else:
+                                c.flags |= USE_MY_METRICS
+                                use_my_metrics_comp = ufo_component_id
+
+                    # We might automatically set the flags if no data is present,
+                    # but:
+                    # - https://github.com/googlefonts/ufo2ft/pull/425 recommends
+                    #   against setting the ROUND_XY_TO_GRID flag
```
Comment (anthrotype) [inline Lib/ufo2ft/instructionCompiler.py:151]: "against"? I re-read that issue and it seems like the consensus was to always set ROUND_XY_TO_GRID.

> for modern rendering environments, the flag doesn't seem to matter; for ClearType it is better to always set it.

File: Lib/ufo2ft/instructionCompiler.py [hunk 13]
```diff
@@ -0,0 +1,194 @@
...
+                    # - USE_MY_METRICS has been set already by
+                    #   outlineCompiler.OutlineTTFCompiler.autoUseMyMetrics
+
+                    if i == 0 and TRUETYPE_OVERLAP_KEY in glyph.lib:
+                        # Set OVERLAP_COMPOUND on the first component only
+                        c.flags &= ~OVERLAP_COMPOUND
```
Comment (anthrotype) [inline Lib/ufo2ft/instructionCompiler.py:157]: why not putting this line in an `else:` block below; feelds slightly weird to me that you first unconditionally un-set it and then set it if True..

File: Lib/ufo2ft/instructionCompiler.py [hunk 14]
```diff
@@ -0,0 +1,194 @@
...
+                table.program = ttLib.tables.ttProgram.Program()
+                table.program.fromAssembly(asm)
+                # Roundtrip once, or if the font is dumped to XML before having
+                # been saved, the assembly code if will look awful.
+                table.program.fromBytecode(table.program.getBytecode())
+
+    def compile_fpgm(self):
+        self._compile_program("fontProgram", "fpgm")
+
+    def compile_glyf(self):
```
Comment (anthrotype) [inline Lib/ufo2ft/instructionCompiler.py:51]: this method is a bit too long, consider splitting it up and factoring out logical chunks of code in distinct private methods maybe

File: Lib/ufo2ft/instructionCompiler.py [hunk 15]
```diff
@@ -0,0 +1,194 @@
...
+        for name in sorted(self.ufo.keys()):
+            glyph = self.ufo[name]
+            ttdata = glyph.lib.get(TRUETYPE_INSTRUCTIONS_KEY, None)
+            if name not in self.font["glyf"]:
+                if ttdata is not None:
+                    logger.warning(
```
Comment (anthrotype) [inline Lib/ufo2ft/instructionCompiler.py:57]: i don't think you want to warn here. It's legitimate that ttf contains less glyphs than the UFO, e.g. if some had been marked as non-export. Just demote this to a debug message if you like to keep

File: Lib/ufo2ft/instructionCompiler.py [hunk 16]
```diff
@@ -0,0 +1,194 @@
...
+                        f"Glyph '{name}' not found in font, "
+                        "skipping compilation of TrueType instructions "
+                        "for this glyph."
+                    )
+                logger.debug(f"UFO keys: {list(self.ufo.keys())}")
```
Comment (anthrotype) [inline Lib/ufo2ft/instructionCompiler.py:62]: printing these (which don't change at each for loop iteration) every time a glyph is missing may be too verbose and noisy, ever for a debug logging message

File: Lib/ufo2ft/instructionCompiler.py [hunk 17]
```diff
@@ -0,0 +1,194 @@
...
+                logger.debug(f"glyf keys: {list(self.font['glyf'].keys())}")
+                continue
+
+            glyf = self.font["glyf"][name]
```
Comment (anthrotype) [inline Lib/ufo2ft/instructionCompiler.py:66]: `glyf` is usually understood as the name of the whole table, not of an individual glyph within the glyf table. `glyf_glyph` or `ttglyph`

File: Lib/ufo2ft/instructionCompiler.py [hunk 18]
```diff
@@ -0,0 +1,194 @@
...
+                        f"Unknown formatVersion {formatVersion} "
+                        "in glyph '{name}', it will have "
+                        "no instructions in font."
+                    )
+                    continue
+
+                # Check if glyph hash matches the current outlines
+                hash_pen = HashPointPen(glyph.width, self.ufo)
+                glyph.drawPoints(hash_pen)
+                glyph_id = ttdata.get("id", None)
```
Comment (anthrotype) [inline Lib/ufo2ft/instructionCompiler.py:80]: is the `id` optional?

File: Lib/ufo2ft/instructionCompiler.py [hunk 19]
```diff
@@ -0,0 +1,194 @@
...
+                # Set component flags
+
+                # We need to decide when to set the flags.
+                # Let's assume if any lib key is not there, or the component
+                # doesn't have an identifier, we should leave the flags alone.
+                use_my_metrics_comp = None
+                for i, c in enumerate(glyf.components):
+                    if i >= len(glyph.components):
+                        logger.error(
+                            "Number of components differ between UFO and TTF "
```
Comment (anthrotype) [inline Lib/ufo2ft/instructionCompiler.py:112]: that'd be weird if it happened, should it not be an error?

File: Lib/ufo2ft/instructionCompiler.py [hunk 20]
```diff
@@ -0,0 +1,194 @@
...
+                            in glyph.lib[OBJECT_LIBS_KEY][ufo_component_id]
+                            or TRUETYPE_METRICS_KEY
+                            in glyph.lib[OBJECT_LIBS_KEY][ufo_component_id]
+                        )
+                    ):
+                        component_lib = glyph.lib[OBJECT_LIBS_KEY][ufo_component_id]
+
+                        c.flags &= ~ROUND_XY_TO_GRID
+                        if component_lib.get(TRUETYPE_ROUND_KEY, False):
+                            c.flags |= ROUND_XY_TO_GRID
```
Comment (anthrotype) [inline Lib/ufo2ft/instructionCompiler.py:134]: i'd prefer an `else:` here but it's probably just a matter of style

File: Lib/ufo2ft/instructionCompiler.py (hunk 10 above)
Comment (jenskutilek) [reply Lib/ufo2ft/instructionCompiler.py:46]: IIRC each PUSH instruction's arguments will be on one line each instead of all on the same line.

File: Lib/ufo2ft/instructionCompiler.py (hunk 11 above)
Comment (jenskutilek) [reply Lib/ufo2ft/instructionCompiler.py:184]: Initially, I had all the hinting compilation stuff in the postProcessor, but a recent optimization (#486) caused the postProcessor to run only on a finished VF, which caused my code to not apply cvt variations. I moved the CVT compilation to the outlineCompiler to build a cvt for each input font.

Running on the VF once is fine for the hinting, because everything except cvt must be identical in all masters. And it should save some processing time.

File: Lib/ufo2ft/instructionCompiler.py (hunk 10 above)
Comment (jenskutilek) [reply Lib/ufo2ft/instructionCompiler.py:46]: I think I noticed that when saving a ttx file at this point while debugging. For normal operation it shouldn't matter. I'll remove it.

File: Lib/ufo2ft/instructionCompiler.py (hunk 9 above)
Comment (jenskutilek) [reply Lib/ufo2ft/instructionCompiler.py:34]: Ok

File: Lib/ufo2ft/instructionCompiler.py (hunk 16 above)
Comment (jenskutilek) [reply Lib/ufo2ft/instructionCompiler.py:63]: I think it can be removed

File: Lib/ufo2ft/instructionCompiler.py (hunk 19 above)
Comment (jenskutilek) [reply Lib/ufo2ft/instructionCompiler.py:113]: It happens when compileTTF is called with `flattenComponents=True`. I'd prefer not to raise an error, but I've changed it so 1) it is tested before setting the component flags, and 2) it skips setting component flags for the glyph completely.

File: Lib/ufo2ft/instructionCompiler.py (hunk 18 above)
Comment (jenskutilek) [reply Lib/ufo2ft/instructionCompiler.py:81]: No. I changed the handling of all required lib keys so they will log an error and skip the whole glyph if missing. (8b115ca40afd513d777e136e5516b151dc9a77c1)

File: Lib/ufo2ft/instructionCompiler.py (hunk 12 above)
Comment (jenskutilek) [reply Lib/ufo2ft/instructionCompiler.py]: Oops ... I probably wrote this before that discussion was finished. I changed the default behaviour to always set ROUND_XY_TO_GRID, and only unset it if explicitly set to false in the lib. (in d8362207d1ff8a830fe72651101f1c3abfc9ad25)

**Comment by jenskutilek** [comment]
I began refactoring of the long `compile_glyf` method. That's it for today, I have to look after my kids now :)

**Comment by jenskutilek** [comment]
Well, this took a while...

I’ve done some more refactoring and cleaning up, I think this is good to go now.

File: Lib/ufo2ft/instructionCompiler.py (hunk 14 above)
Comment (jenskutilek) [reply Lib/ufo2ft/instructionCompiler.py:177]: Done.

File: Lib/ufo2ft/instructionCompiler.py (hunk 11 above)
Comment (jenskutilek) [reply Lib/ufo2ft/instructionCompiler.py]: I ran some comparisons with compiling in the outlineCompiler vs. postProcessor:

No hinting: 43.074s
Compilation in outlineCompiler: 43.797s
Compilation in postProcessor: 43.597s

(Thats the complete processing time to build a VF from 4 master UFOs, 1500 glyphs)

No big difference there. It may be a bigger difference when there is a large number of masters, that’s why I tend to keep the hinting compilation in postProcessor.

The additional font compilation/reload time for the "hack" is about 380 ms for a 1500 glyph variable font. The raw hinting compile time is about 280 ms per master on my machine. So for more than 2 masters, the compilation in postProcessor should reduce the total processing time if those numbers scale.

**Comment by jenskutilek** [comment]
If there are no objections, I will merge this in a couple of days.

**Comment by anthrotype** [comment]
> If there are no objections, I will merge this in a couple of days.

hey, sorry Jens, I was on leave and couldn't take a look. I hope to be able to do that soon, ideally before a new release.

**Comment by anthrotype** [comment]
If I may, I suggest the next time we merge a major change like this one into the main branch, one gets at least another collaborator to review it, as it's customary. In general I prefer we try do this voluntarily instead of having Github enforce this rule automatically for any PR, however minor. Thanks

**Comment by anthrotype** [comment]
Sure, no worries. And thanks again for putting so much effort into this, and apologies for the slow response on my part.

File: Lib/ufo2ft/outlineCompiler.py [hunk 21]
```diff
@@ -118,6 +121,7 @@ def __init__(
         self._glyphBoundingBoxes = None
         self._fontBoundingBox = None
         self._compiledGlyphs = None
+        self.compileTrueTypeHinting = compileTrueTypeHinting
```
Comment (anthrotype) [inline Lib/ufo2ft/outlineCompiler.py:124]: I would expect `compileTrueTypeHinting` to only be available for the TrueType outline compiler, not for the base compiler class.

File: Lib/ufo2ft/outlineCompiler.py [hunk 22]
```diff
@@ -1460,6 +1464,8 @@ def setupOtherTables(self):
         self.setupTable_glyf()
         if self.ufo.info.openTypeGaspRangeRecords:
             self.setupTable_gasp()
+        if self.compileTrueTypeHinting:
+            self.setupTable_cvt()
```
Comment (anthrotype) [inline Lib/ufo2ft/outlineCompiler.py:1468]: it's confusing to read an option called compileTrueTypeHinting whose effect is to only compiling the CVT table and not the rest of the hinting tables. I'd remove the option altogether. If there are TRUETYPE_INSTRUCTIONS_KEY in the ufo you build whatever you are requested to build, otherwise you don't.

File: Lib/ufo2ft/postProcessor.py [hunk 23]
```diff
@@ -98,6 +99,13 @@ def process(
           is True and CFF or CFF2 table is present. Choose between "cffsubr" or
           "compreffor". By default "cffsubr" is used for both CFF 1 and CFF 2.
           NOTE: compreffor currently doesn't support input fonts with CFF2 table.
+
+        compileTrueTypeHinting:
+          Compile TrueType hinting from the UFO, if present. You must make sure
+          that the outlines from the UFO have not been altered before this
+          step. The point indices change e.g. when removing overlaps or
+          changing the path directions. This will most certainly make any
+          hinting data in the source UFO invalid.
```
Comment (anthrotype) [inline Lib/ufo2ft/postProcessor.py:108]: if that's the case, then we should probably warn somewhere if both compileTrueTypeHinting and other incompatible options are set

File: Lib/ufo2ft/postProcessor.py (hunk 2 above)
Comment (anthrotype) [reply Lib/ufo2ft/postProcessor.py:122]: I really don't like this recompilation step, we need to go to the bottom of the issue and try to avoid this.

**Review by anthrotype — COMMENTED** [review]
sorry but I'm going to revert this PR for now, there's a few things that I'd like to fix before we can ship it. I left a few comments below. Besides I don't like you're using the integration_test.py to test this, with so many new UFO and ttx test files added that nobody (except the author) will ever bother to review individually, with the tests simply ratifying with a ttx diff whatever comes out of the compiler. We should have more tailored unit tests for each specific functionality added, with test data reduced to a minimum, ideally generated by code and placed near to where the test code is so it's clear to the reader what it's actually testing.  
I'll push it to a branch until I find the time to clean things up.

File: Lib/ufo2ft/outlineCompiler.py (hunk 22 above)
Comment (jenskutilek) [reply Lib/ufo2ft/outlineCompiler.py:1468]: I agree and changed the behaviour in #577
